### ISSUE TITLE
xs:sequence encoded out of order

### ISSUE THREAD

**Issue by pythonator** [brunato/xmlschema#69]
Hello there,
Thanks for this great package! I saw that you started working on encoding lately so I am reporting a bug.

The children of `xs:sequence` do not appear in the output in the order defined by the sequence in the XSD.

Here is an example:
```python
import xmlschema
import xml.etree

XSD_PATH = r"C:\tmp\xmlschema\shiporder.xsd"
XML_PATH = r"C:\tmp\xmlschema\shiporder.xml"

def reordered_dict(d):
    """This function is needed to cause the bug to happen in Python 3.6+ due to dict being ordered."""
    return {k: d[k] for k in sorted(d)}


xsd = xmlschema.XMLSchema(XSD_PATH)
xsd.validate(XML_PATH)
d = xsd.to_dict(XML_PATH)
for bad_attr in ('@xsi:schemaLocation', '@xsi:noNamespaceSchemaLocation'):
    if bad_attr in d:
        del d[bad_attr]
new_etree = xsd.encode(d)
new_xml_text = xml.etree.ElementTree.tostring(new_etree).decode('us-ascii')
print(new_xml_text)
xsd.validate(new_xml_text)

rd = reordered_dict(d)
rnew_etree = xsd.encode(rd)
rnew_xml_text = xml.etree.ElementTree.tostring(rnew_etree).decode('us-ascii')
print(rnew_xml_text)
import time; time.sleep(0.01)  # Separate stdout and stderr in PyCharm
xsd.validate(rnew_xml_text)
```

Where XSD_PATH and XML_PATH point to files taken from [here](https://www.w3schools.com/xml/schema_example.asp).

In python 3.6+, the second `xsd.validate(new_xml_text)` will fail. In previous Python versions, the first validate will probably fail.

**Comment by brunato** [comment]
Thank you for reporting about this problem.

I'm working heavily on encoding to close the 1.0 release (the first release that supports XSD with all the features previewed for this library).

For Python < 3.6 the `collections.OrderedDict` is needed for avoiding this issue. I just wrote for next version two helper API at module level for handling the decoding to JSON and from JSON in a simpler way, but for the decode/encode API you need to pass the optional argument `dict_class=OrderedDict`.

**Comment by pythonator** [comment]
I think using an ordered dict is only a workaround.

What if I want to

* Load a XML into a dict.
* Add an optional tag into the middle of a sequence
* Save to XML

This isn't possible using an ordered dict.

Preferably, I could use normal dicts and not bother about order. `encode` would order the tags based on the sequence definition in the XSD instead of their order in the dict.

**Comment by brunato** [comment]
The ordering of items is important to strictly validate the data. The work on encoding is now around this goal, that it's not achieved yet for model groups (`iter_elements()` is not strictly model conforming). Anyway I think there could be space in the future for a different approach with _lax_ or _skip_ validation modes, always in the method `XsdGroup.iter_encode()`, to rebuild an XML from unordered data. 

Anyway from Python 3.6 in CPython also the `dict()` objects preserve insertion order like `collection.OrderedDict` and more recently "[Dict keeps insertion order](https://mail.python.org/pipermail/python-dev/2017-December/151266.html)" has become a rule of the language.

**Comment by brunato** [comment]
Encoding improved in 1.0.1 and 1.0.2 releases but additional work is needed for this.

**Comment by brunato** [comment]
A new model validation method, based on a model visitor class, has been implemented in version 1.0.3.   
This new version should reorders children in case of model errors if validation is not 'strict'. You could be try it and reports if it works.

**Comment by brunato** [comment]
No response for a while. Assumed that is resolved.
